### PR TITLE
Project repository functionality update

### PR DESCRIPTION
**Pull request by Yumfriez**

### CODE REVIEW

File: src/main/java/com/epam/ta/reportportal/dao/LaunchRepositoryCustom.java [hunk 1]
```diff
@@ -71,4 +73,6 @@
 	 * @return Map of username -> launches count
 	 */
 	Map<String, Integer> countLaunchesGroupedByOwner(Long projectId, String mode, LocalDateTime from);
+
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/LaunchRepositoryCustom.java]: java docs for interface methods are needed

File: src/main/java/com/epam/ta/reportportal/dao/ProjectRepositoryCustomImpl.java [hunk 2]
```diff
@@ -132,4 +136,29 @@ public ProjectInfo findProjectInfoFromDate(Long projectId, LocalDateTime fromDat
 	public List<String> findAllProjectNames() {
 		return dsl.select(PROJECT.NAME).from(PROJECT).fetchInto(String.class);
 	}
+
+	@Override
+	public Page<Project> findAllIdsAndProjectAttributes(ProjectAttributeEnum projectAttribute, Pageable pageable) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/ProjectRepositoryCustomImpl.java]: could be done using filters after the new query builder

File: src/main/java/com/epam/ta/reportportal/entity/enums/KeepLaunchDelay.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	public long getDays() {
+		return days;
+	}
+
+	KeepLaunchDelay(String delay, long period) {
+		this.value = delay;
+		this.days = period;
+	}
+
+	public static KeepLaunchDelay getByName(String type) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/entity/enums/KeepLaunchDelay.java]: remove this method

File: src/main/java/com/epam/ta/reportportal/commons/querygen/constant/ProjectCriteriaConstant.java [hunk 4]
```diff
@@ -22,4 +22,5 @@
 public class ProjectCriteriaConstant {
 	public static final String CRITERIA_PROJECT_TYPE = "type";
 	public static final String CRITERIA_PROJECT_NAME = "name";
+	public static final String CRITERIA_ATTRIBUTE_NAME = "attribute_name";
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/commons/querygen/constant/ProjectCriteriaConstant.java]: camel case

File: src/main/java/com/epam/ta/reportportal/dao/LogRepositoryCustomImpl.java [hunk 5]
```diff
@@ -131,4 +154,35 @@ public Integer getPageNumber(Long id, Filter filter, Pageable pageable) {
 
 		return BigDecimal.valueOf(rowNumber).divide(BigDecimal.valueOf(pageable.getPageSize()), RoundingMode.CEILING).intValue();
 	}
+
+	@Override
+	public boolean hasLogsAddedLately(Duration period, Long launchId, StatusEnum... statuses) {
+		List<JStatusEnum> jStatuses = Arrays.stream(statuses).map(it -> JStatusEnum.valueOf(it.name())).collect(toList());
+		Timestamp before = Timestamp.from(Instant.now().minusSeconds(period.getSeconds()));
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/LogRepositoryCustomImpl.java]: i can see similar logic of counting before timestamp in several methods. could you please make one common method. it exists in a previous version of rp. find it and implement something similar here

File: src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java [hunk 6]
```diff
@@ -122,6 +129,34 @@ public Boolean hasItemsInStatusByParent(Long parentId, StatusEnum... statuses) {
...
+	@Override
+	public Boolean hasItemsInStatusAddedLately(Long launchId, Duration period, StatusEnum... statuses) {
+		List<JStatusEnum> jStatuses = Arrays.stream(statuses).map(it -> JStatusEnum.valueOf(it.name())).collect(toList());
+		return dsl.fetchExists(dsl.selectOne()
+				.from(TEST_ITEM)
+				.join(TEST_ITEM_RESULTS)
+				.onKey()
+				.where(TEST_ITEM.LAUNCH_ID.eq(launchId))
+				.and(TEST_ITEM_RESULTS.STATUS.in(jStatuses))
+				.and(TEST_ITEM.LAST_MODIFIED.gt(Timestamp.from(Instant.now().minusSeconds(period.getSeconds()))))
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java]: here

File: src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java [hunk 7]
```diff
@@ -122,6 +129,34 @@ public Boolean hasItemsInStatusByParent(Long parentId, StatusEnum... statuses) {
...
+		List<JStatusEnum> jStatuses = Arrays.stream(statuses).map(it -> JStatusEnum.valueOf(it.name())).collect(toList());
+		return dsl.fetchExists(dsl.selectOne()
+				.from(TEST_ITEM)
+				.join(TEST_ITEM_RESULTS)
+				.onKey()
+				.join(LOG)
+				.on(TEST_ITEM.ITEM_ID.eq(LOG.ITEM_ID))
+				.where(TEST_ITEM.LAUNCH_ID.eq(launchId))
+				.and(TEST_ITEM_RESULTS.STATUS.in(jStatuses))
+				.and(TEST_ITEM.LAST_MODIFIED.lt(Timestamp.from(Instant.now().minusSeconds(period.getSeconds()))))
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java]: here

File: src/main/java/com/epam/ta/reportportal/dao/LaunchRepositoryCustom.java (hunk 1 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/LaunchRepositoryCustom.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/ProjectRepositoryCustomImpl.java (hunk 2 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/ProjectRepositoryCustomImpl.java]: done

File: src/main/java/com/epam/ta/reportportal/commons/querygen/constant/ProjectCriteriaConstant.java (hunk 4 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/commons/querygen/constant/ProjectCriteriaConstant.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/LogRepositoryCustomImpl.java (hunk 5 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/LogRepositoryCustomImpl.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java (hunk 6 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java (hunk 7 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/TestItemRepositoryCustomImpl.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package com.epam.ta.reportportal.dao.util;
+
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * @author <a href="mailto:ivan_budayeu@epam.com">Ivan Budayeu</a>
+ */
+public final class TimeCalculatorUtil {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java]: Let's make a clear name. Something like DateUtils or TimestampUtils

File: src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+	private TimeCalculatorUtil() {
+		//static only
+	}
+
+	public static Timestamp getTimeBeforeNowWithPeriod(Duration period) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java]: probably it's better to give a more clear name for the method. some kind of:
getTimestampBackFromNow(Duration period)

File: src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java (hunk 8 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java (hunk 9 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/util/TimeCalculatorUtil.java]: done

**Pull request merged**